### DISCUSSION TITLE
Waiting on a file descriptor without blocking the event loop

### DISCUSSION THREAD

**Discussion by bitcompost** [janet-lang/janet#1320] [Q&A] [answered]
Hi, I'm writing a small gui library experiment in janet and C, and on the C side I connect to wayland compositor using wl_display_connect() and then get the corresponding fd with wl_display_get_fd() and need to call wl_display_dispatch() whenever its possible to read from the fd. 
My current hacky solution is to do a nonblocking poll in C and if no data incoming then call (ev/sleep 0.001) to let the other fibers run, but that takes like 5% CPU when doing nothing.
Another idea I had is to do the poll in another thread but that seems like unnecessary complication.
Is there a better way to do this?

**Comment by bitcompost** [comment]
<s>I think I figured it out</s> by reading the Janet event loop implementation.</s> I created a stream by doing net/connect on the wayland socket, then on the C side I unwrapped the stream abstract type and got the fd from it, then I used function which wayland luckily has: "wl_display_connect_to_fd()" to setup the connection. Now I can wait on the fd by reading zero bytes from the stream: (ev/read stream 0).

**Reply by bitcompost** [reply]
Well, this also probably wasn't an ideal solution, because it stopped working after I updated to newest Janet (master branch). Now (ev/read stream 0) doesn't block.

**Reply by bitcompost** [reply]
Also, there's a bug. After a stream is created ev/sleep takes 100% CPU. Steps to reproduce:
```
Janet 1.32.1-meson linux/aarch64/gcc - '(doc)' for help
repl:1:> (net/connect "1.1.1.1" "80")
<core/stream 0xAAAACB3E0010>
repl:2:> (ev/sleep 42)
```
https://github.com/janet-lang/janet/issues/1321

**Reply by bakpakin** [reply]
Your posted example does not cause increased cpu usage on my machine. Is there any other setup you did to get the above code to repro with 100% CPU usage?

Also yes, the event loop implementation changed to make fewer system calls. We no longer call epoll_ctl for each read and each write.

**Comment by bakpakin** [comment, marked as answer]
As far as the original issue (100% cpu issue aside), the new code to do this in C is as follows:

```c

static void callback(JanetFiber *fiber, JanetAsyncEvent event) {
    switch(event) {
        default:
            return;
        case JANET_ASYNC_EVENT_READ:
        case JANET_ASYNC_EVENT_HUP:
        case JANET_ASYNC_EVENT_ERR:
            break;
    }
    janet_schedule(fiber, janet_wrap_nil());
    janet_async_end(fiber);
}

JANET_FN(cfun_wayland_wait,
        "(wayland/wait stream)",
        "Wait a stream to be ready for reading, but do not actually read from the stream.") {
    janet_fixarity(argc, 1);
    JanetStream *stream = janet_getabstract(argv, 0, &janet_stream_type);
    // Optionally do an early check here to see if the file descriptor is ready using a single call to poll or similar.
    janet_async_start(stream, JANET_ASYNC_LISTEN_READ, callback, NULL);
}
```

This is essentially what `net/read` and `ev/read` do except without the actual read.

Then, in your script, it would look something like:

```janet
# Get my stream somehow using wl_display_get_fd
(def my-stream (wayland/connect ...))
(wayland/wait my-stream)
(wayland/display-dispatch)
```

**Reply by bitcompost** [reply]
Thanks, it works! And I can detect a lost connection in the callback, really cool. Is this API going to change in the future?

**Reply by bakpakin** [reply]
No, this is the new event loop API that is a refined version of the old JanetListener API.

This functionality is not built into the core library as there isn't a good analog for this on windows.
